### PR TITLE
[Refactoring] Improve IMAP Search command parser code quality

### PR DESCRIPTION
**Pull request by chibenwa**
Rely on String switch case to greatly improve readability.

This changeset is unit tested (I know, that's unexpected).

### CODE REVIEW

**Comment by chibenwa** [comment]
Thanks!

Well that is now way easy to review compare to what it was before!

**Comment by mbaechler** [comment]
> hard to review, but +1 at first sight

split view + ignore whitespace makes it doable

**Comment by chibenwa** [comment]
> The testsuite is not good.

Then that's a good opportunity to improve it.

> Also, you broke what looks like a try to write a very optimized parser [...] 

Point 1: I trust java compiler and java core devs to build a similar character matching tree, whithout us paying the price in the code. Sure some micro benchmark might be usefull here but given the expected length of the input (SEARCH request) I would not expect it to be critical.

> With request.atom(), If I send a very long string as input, it will break the server.

You got a point. Well we could easily have a `MaximumSizeCharValidator` for preventing this to happen.

> However, it's yet another reason to stop writing parsers by hand as if it's still 1980.

:+1: but writing such parboil stuff is way more complicated, and may come in a way further timerange than this actual improvment, that allows the mere average human I am to actually understand (and maintain!) what is going on here.

Hence despite its defaults I would plead for this changeset to be overall valuable.

File: protocols/imap/src/main/java/org/apache/james/imap/decode/parser/SearchCommandParser.java [hunk 1]
```diff
@@ -74,68 +75,97 @@ protected SearchKey searchKey(ImapSession session, ImapRequestLineReader request
...
-                return f(request, charset);
-            case 'G':
-                throw new DecodingException(HumanReadableText.ILLEGAL_ARGUMENTS, "Unknown search key");
-            case 'H':
+            return parseSearchKey(session, request, charset, isFirstToken);
+        }
+    }
+
+    private SearchKey parseSearchKey(ImapSession session, ImapRequestLineReader request, Charset charset, boolean isFirstToken) throws DecodingException {
+        String searchKey = request.atom().toUpperCase(Locale.US);
```
Comment (remk) [inline protocols/imap/src/main/java/org/apache/james/imap/decode/parser/SearchCommandParser.java]: To mitigate the issue that @mbaechler noticed, you can get only the first X characters ( where X is the size of the biggest key name) and stop at the first blank character.

Comment (chibenwa) [reply protocols/imap/src/main/java/org/apache/james/imap/decode/parser/SearchCommandParser.java]: Yes, that is my MaximumSize char validator proposal.

**Comment by mbaechler** [comment]
> > The testsuite is not good.
> 
> Then that's a good opportunity to improve it.

:+1: 

> > Also, you broke what looks like a try to write a very optimized parser [...]
> 
> Point 1: I trust java compiler and java core devs to build a similar character matching tree, whithout us paying the price in the code. Sure some micro benchmark might be usefull here but given the expected length of the input (SEARCH request) I would not expect it to be critical.

Well, I agree that defaults are decent for most code but when you are talking about the core of your business (is IMAP command parsing our core business?), it may not be enough.

I would not request such optimization if it were not a regression from the work of somebody that did care to optimize it in the past.

> > With request.atom(), If I send a very long string as input, it will break the server.
> 
> You got a point. Well we could easily have a `MaximumSizeCharValidator` for preventing this to happen.

I can be a solution, yes.

> > However, it's yet another reason to stop writing parsers by hand as if it's still 1980.
> 
> +1 but writing such parboil stuff is way more complicated, and may come in a way further timerange than this actual improvment, that allows the mere average human I am to actually understand (and maintain!) what is going on here.

I don't agree that parpoiled (or any grammar parser) is harder to understand as it's quite the opposite: it makes the grammar very readable because it doesn't describe "how to parse" but "what to parse".

However It requires to learn a new skill.

It may not be yet in the toolbox of all the team members but as a mail server is very much about parsing, I expect that we invest in this skill.

> Hence despite its defaults I would plead for this changeset to be overall valuable.

It agree if we protect ourselves against a very long atom.

**Comment by openpaas-james** [comment]
Build started sha1 is merged.

**Comment by chibenwa** [comment]
I won't continue that work, it has been rewritten in scala. Abort.

**Pull request closed without merging**